alf/algorithms/algorithm.py

```diff
@@ -1533,6 +1714,214 @@ def _update(self, experience, batch_info, weight):
+                                                   experience)
+                    if batch_info:
+                        binfo = alf.nest.map_structure(
+                            lambda x: x[batch_indices] if isinstance(
+                                x, torch.Tensor) else x, batch_info)
+                    else:
+                        binfo = None
+                    batch = _make_time_major(batch)
+                    batch = alf.nest.utils.convert_device(batch)
+                else:
+                    batch = None
+                    binfo = None
+
+                offline_batch = alf.nest.map_structure(
+                    lambda x: x[batch_indices], offline_experience)
+                if offline_batch_info:
+                    offline_binfo = alf.nest.map_structure(
+                        lambda x: x[batch_indices] if isinstance(
+                            x, torch.Tensor) else x, offline_batch_info)
+                else:
+                    offline_binfo = None
+                offline_batch = _make_time_major(offline_batch)
+                offline_batch = alf.nest.utils.convert_device(offline_batch)
+
+                (exp, train_info, loss_info, offline_exp, offline_train_info,
+                 offline_loss_info, params) = self._hybrid_update(
+                     batch,
+                     binfo,
+                     offline_batch,
+                     offline_binfo,
+                     weight=alf.nest.get_nest_size(offline_batch, 1) /
+                     mini_batch_size)
+                if do_summary:
+                    if exp:
+                        self.summarize_train(exp, train_info, loss_info,
+                                             params)
+                    # TODO: separate summarization
+                    self.summarize_train(offline_exp, offline_train_info,
+                                         offline_loss_info, None)
+
+        train_steps = 2 * batch_size * mini_batch_length * num_updates
+        return train_steps
+
+    def _hybrid_update(self, experience, batch_info, offline_experience,
+                       offline_batch_info, weight):
+        length = alf.nest.get_nest_size(offline_experience, dim=0)
+
+        if self._count == 0 or (
+            (self._count > self._rl_train_after_update_steps) and
+            (self._count % self._rl_train_every_update_steps == 0)):
+            # 1) RL training
+            RL_train = True
+            pre_train = False
+            if self._config.temporally_independent_train_step or length == 1:
+                train_info = self._collect_train_info_parallelly(experience)
+            else:
+                train_info = self._collect_train_info_sequentially(experience)
+
+            experience = dist_utils.params_to_distributions(
+                experience, self.processed_experience_spec)
+
+            loss_info = self.calc_loss(train_info)
```

Comment:
use `_collect_train_info_and_loss_info()`

Reply:
changed to using this function